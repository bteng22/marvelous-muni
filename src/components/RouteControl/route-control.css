.route-control__container {
  display: flex;
  margin: 20px;
  flex-direction: column;
  width: 250px;
  height: 350px;
  position: absolute;
  right: 0;
  overflow-y: scroll;
  border-radius: 3px;
}

.route-control__button {
  display: none;
}

.route-control__close {
  display: none;
}

.route-list {
  padding: 0;
  margin: 0;
  list-style-type: none;
}

.route-list-item {
  display: flex;
  background: #f2f2f2;
  border: 1px solid #000;
}

.route-list-item > input {
  margin: 10px;
}

.route-list-item > span {
  display: inline-flex;
  align-items: center;
  width: 100%;
  padding-left: 10px;
  border-left: 1px solid #000;
}

.route-list-item__title > span {
  display: inline-flex;
  justify-content: center;
  padding: 7px;
  margin-left: 32px;
  font-weight: 700;
  font-size: 20px;
}

@media (max-width: 991px) {
  .route-control__container {
    height: 100%;
    margin: 0;
  }

  .route-list {
    transition: transform 0.5s;
    transform: translateX(250px)
  }

  .route-list.slide {
    transform: translateX(0);
  }

  .route-control__button {
    display: initial;
    margin-top: 20px;
    position: absolute;
    right: 0;
  }
  
  .route-control--exit {
    margin-top: 0px;
    position: initial;
    right: initial;
    align-self: center;
    font-style: normal;
    padding: 5px;
    cursor: pointer;
  }
}