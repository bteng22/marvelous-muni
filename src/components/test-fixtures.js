export const fakeRouteList = {
  '44': {
    title: "44-O'Shaughnessy",
    tag: "44",
    visible: true
  },
  "KT": {
    title: "KT-Ingleside/Third Street",
    tag: "KT",
    visible: true
  }
};

export const fakeVehicleData = [
  {
    dirTag: "44___I_F00",
    heading: "218",
    id: "8811",
    lat: "37.712856",
    lon: "-122.433144",
    predictable: "true",
    routeTag: "44",
    secsSinceReport: "40",
    speedKmHr: "0"
  },
  {
    dirTag: "KT___I_F00",
    heading: "135",
    id: "1538",
    lat: "37.790501",
    lon: "-122.389717",
    predictable: "true",
    routeTag: "KT",
    secsSinceReport: "17",
    speedKmHr: "16"
  }
];