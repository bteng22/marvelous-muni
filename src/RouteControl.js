import React, { Component } from 'react';
import './route-control.css';

class RouteControl extends Component {
  constructor(props) {
    super(props);
    this.state = {
      routeList: this.props.appState.routeList
    }
    this.fetchRouteList = this.fetchRouteList.bind(this);
  }

  fetchRouteList() {
    fetch(`http://webservices.nextbus.com/service/publicJSONFeed?command=routeList&a=sf-muni`)
      .then(res => res.json())
      .then((json) => {
        const initialRoutes = json.route.reduce((routes, route) => {
          routes[route.tag] = {
            ...route,
            visible: true            
          }
          return routes;
        }, {});
        this.props.setAppState({
          routeList: initialRoutes
        })
      })
  }

  toggleRouteCheckbox(route) {
    const { routeList } = this.props.appState;
    const updatedRoutes = Object.assign({}, routeList, {
      [route.tag]: {
        ...route,
        visible: !route.visible
      }
    })

    this.props.setAppState({
      routeList: updatedRoutes
    })
  }

  renderRouteList(routeList) {
    return routeList.map((route) => {
      return (
        <li key={route.tag}>
          <input
            type="checkbox"
            checked={route.visible}
            onChange={() => this.toggleRouteCheckbox(route)}
          />{route.title}
        </li>
      )
    })
  }
  
  componentDidMount() {
    this.fetchRouteList() 
  }

  render() {
    const { routeList } = this.props.appState;
    return (
      <div className='route-control__container'>
        <ul className='route-list'>
          { this.renderRouteList(Object.values(routeList))}
        </ul> 
      </div>
    );
  }
}

export default RouteControl;
