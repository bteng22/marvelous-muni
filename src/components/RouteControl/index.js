import React, { Component, Fragment } from 'react';
import './route-control.css';

class RouteControl extends Component {
  constructor(props) {
    super(props);
    this.state = {
      routeList: this.props.appState.routeList,
      showMenu: false
    }
    this.fetchRouteList = this.fetchRouteList.bind(this);
    this.toggleAllRoutes = this.toggleAllRoutes.bind(this);
    this.toggleRouteMenu = this.toggleRouteMenu.bind(this);
  }

  fetchRouteList() {
    fetch(`http://webservices.nextbus.com/service/publicJSONFeed?command=routeList&a=sf-muni`)
      .then(res => res.json())
      .then((json) => {
        const initialRoutes = json.route.reduce((routes, route) => {
          routes[route.tag] = {
            ...route,
            visible: true            
          }
          return routes;
        }, {});
        this.props.setAppState({
          routeList: initialRoutes
        })
      })
  }

  toggleAllRoutes(event) {
    const { routeList } = this.props.appState;
    const routeListCopy = Object.assign({}, routeList);
    Object.keys(routeList).forEach((routeTag) => {
      routeListCopy[routeTag].visible = event.target.checked;
    });

    this.props.setAppState({
      routeList: routeListCopy
    });
  }

  toggleRouteCheckbox(route) {
    const { routeList } = this.props.appState;
    const updatedRoutes = Object.assign({}, routeList, {
      [route.tag]: {
        ...route,
        visible: !route.visible
      }
    })

    this.props.setAppState({
      routeList: updatedRoutes
    })
  }

  toggleRouteMenu() {
    this.setState({
      showMenu: !this.state.showMenu
    });
  }

  renderRouteList(routeList) {
    return routeList.map((route) => {
      return (
        <li 
          key={ route.tag }
          className="route-list-item"
        >
          <input
            type="checkbox"
            checked={route.visible}
            onChange={() => this.toggleRouteCheckbox(route)}
          />
          <span>{ route.title }</span>
        </li>
      )
    })
  }
  
  componentDidMount() {
    this.fetchRouteList() 
  }

  render() {
    const { routeList } = this.props.appState;
    const { showMenu } = this.state;
    return (
      <Fragment>
        <div className='route-control__container'>
          <button
            className='route-control__button'
            onClick={this.toggleRouteMenu}
          >
            Route Toggles >
          </button>
          <ul className={`route-list ${showMenu ? 'slide' : ''}`}>
            <li className="route-list-item route-list-item__title">
              <span>Route Toggles</span>
              <i
                onClick={this.toggleRouteMenu}
                className="route-control__button route-control--exit"
              >
                X
              </i>
            </li>
            <li className="route-list-item">
              <input
                type="checkbox"
                onChange={this.toggleAllRoutes}  
              />
              <span>Select All</span>
            </li>
            { this.renderRouteList(Object.values(routeList))}
          </ul> 
        </div>
      </Fragment>
    );
  }
}

export default RouteControl;
