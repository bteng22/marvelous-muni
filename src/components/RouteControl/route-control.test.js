import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import RouteControl from './index';
import { fakeRouteList } from '../test-fixtures';

describe('RouteControl', () => {
  const promise = Promise.resolve({ 
    json: () => Promise.resolve({ route: [{ title: '44 - Bus', tag: '44'}] })
  })
  let props;

  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() => promise);
    props = {
      appState: {
        routeList: fakeRouteList
      },
      setAppState: jest.fn()
    }
  });

  it('should render successfully', () => {
    shallow(<RouteControl {...props} />, {
      disableLifecycleMethods: true
    })
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(<RouteControl {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should toggle route checkbox', () => {
    const route = {
      tag: '44',
      visible: true
    };
    const wrapper = shallow(<RouteControl {...props} />);
    wrapper.instance().toggleRouteCheckbox(route);
    expect(props.setAppState).toHaveBeenCalledWith({"routeList": {"44": {"tag": "44", "visible": false}, "KT": {"tag": "KT", "title": "KT-Ingleside/Third Street", "visible": true}}})
    route.visible = false;
    wrapper.instance().toggleRouteCheckbox(route);
    expect(props.setAppState).toHaveBeenCalledWith({"routeList": {"44": {"tag": "44", "visible": true}, "KT": {"tag": "KT", "title": "KT-Ingleside/Third Street", "visible": true}}})
  });

  it('should toggle all routes', () => {
    const event = {
      target: { checked: false }
    };
    const wrapper = shallow(<RouteControl {...props} />);
    expect(props.appState.routeList).toEqual({"44": {"tag": "44", "title": "44-O'Shaughnessy", "visible": true}, "KT": {"tag": "KT", "title": "KT-Ingleside/Third Street", "visible": true}});
    wrapper.instance().toggleAllRoutes(event);
    expect(props.setAppState).toHaveBeenCalledWith({"routeList": {"44": {"tag": "44", "title": "44-O'Shaughnessy", "visible": false}, "KT": {"tag": "KT", "title": "KT-Ingleside/Third Street", "visible": false}}});
  });

  it('should fetch route list', () => {
    const wrapper = shallow(<RouteControl {...props} />, {
      disableLifecycleMethods: false
    });

    return promise.then().then(() => {
      expect(global.fetch).toHaveBeenCalledWith('http://webservices.nextbus.com/service/publicJSONFeed?command=routeList&a=sf-muni');
      expect(props.setAppState).toHaveBeenCalledWith({
        routeList: {
          44: {
            tag: "44",
            title: "44 - Bus",
            visible: true
          }
        }
      });
    })
  });

  it('should toggle route menu', () => {
    const wrapper = shallow(<RouteControl {...props} />);
    expect(wrapper.state('showMenu')).toBe(false);
    wrapper.instance().toggleRouteMenu();
    expect(wrapper.state('showMenu')).toBe(true);
  });

  it('should apply slide class when showMenu is true', () => {
    const wrapper = shallow(<RouteControl {...props} />);
    expect(wrapper.find('.route-list').hasClass('slide')).toBe(false);
    wrapper.setState({
      showMenu: true
    });
    wrapper.update()
    expect(wrapper.find('.route-list').hasClass('slide')).toBe(true);
  });
});