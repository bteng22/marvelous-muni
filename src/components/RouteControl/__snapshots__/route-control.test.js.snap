// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteControl should match snapshot 1`] = `
<div
  className="route-control__container"
>
  <button
    className="route-control__button"
    onClick={[Function]}
  >
    Route Toggles &gt;
  </button>
  <ul
    className="route-list "
  >
    <li
      className="route-list-item route-list-item__title"
    >
      <span>
        Route Toggles
      </span>
      <i
        className="route-control__button route-control--exit"
        onClick={[Function]}
      >
        X
      </i>
    </li>
    <li
      className="route-list-item"
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        Select All
      </span>
    </li>
    <li
      className="route-list-item"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        44-O'Shaughnessy
      </span>
    </li>
    <li
      className="route-list-item"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        KT-Ingleside/Third Street
      </span>
    </li>
  </ul>
</div>
`;
